sound/soc/intel/boards/sof_es8336.c

```diff
@@ -111,6 +124,45 @@ static void log_quirks(struct device *dev)
 		dev_info(dev, "quirk headset at mic1 port enabled\n");
 }
 
+static void pcm_pop_work_events(struct work_struct *work)
+{
+	struct sof_es8336_private *priv =
+		container_of(work, struct sof_es8336_private, pcm_pop_work.work);
+
+	gpiod_set_value_cansleep(priv->gpio_speakers, priv->speaker_en);
+
+	if (quirk & SOF_ES8336_HEADPHONE_GPIO)
+		gpiod_set_value_cansleep(priv->gpio_headphone, priv->speaker_en);
+				
+}
+
+static int sof_8336_trigger(struct snd_pcm_substream *substream, int cmd)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_card *card = rtd->card;
+	struct sof_es8336_private *priv = snd_soc_card_get_drvdata(card);
+	
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_START:
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+	case SNDRV_PCM_TRIGGER_RESUME:
+		break;
+
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_STOP:
+		if(priv->speaker_en == false)
+			if(substream->stream == 0)
+				gpiod_set_value_cansleep(priv->gpio_speakers, true);		
```

Comment:
no idea what you are trying to do here.

Reply:
The speaker need to be turned off before dapm turned off, so they are turned off here. Is there any better way to do this?

Reply:
can you clarify what level enables the speakers? Here you set the GPIO line to high, in other places you set it to speaker_en. Not clear at all

Reply:
The speaker is turned on by set it's value to false `gpiod_set_value_cansleep(priv->gpio_speakers, false);`
Because an 'active_low'  gpio is requested, and an 'active_low'  gpio turned off means high level, which turns on the speaker.

Reply:
ok, it's really confusing that you are using speaker_en in a inverted way, this is far from self-explanatory.